### ISSUE TITLE
Error Running GiniClust

### ISSUE THREAD

**Issue by nlawlor** [lanjiangboston/GiniClust#1]
Hello,

I attempted to run GiniClust as an R script using the command line:

Rscript GiniClust_Main.R -f expression.data.csv -t RNA-seq -o expression_results

However, I receive the following error:

"
Error in FUN(newX[, i], ...) : invalid 'type' (character) of argument
Calls: GiniClust_Preprocess -> apply
Execution halted
"

I believe this error message is suggesting my expression matrix is wrongly formatted. My matrix is formatted with genes as the row names and samples as the column names. 

How should the input expression CSV file be formatted to run GiniClust properly? 

When I run GiniClust using the sample dataset (Data_GBM.csv) and the command:

 Rscript GiniClust_Main.R -f Data_GBM.csv -t RNA-seq -o GBM_results

I do not receive an error message. When I looked at the format of the Data_GBM.csv file, I noticed that the file was in the same format of genes as row names and sample as column names. 

Thank you,
Nathan

All required packages to run GiniClust were successfully loaded as shown below:

Attaching package: ‘zoo’

The following objects are masked from ‘package:base’:

```
as.Date, as.Date.numeric
```

Loaded glmnet 2.0-5

Attaching package: ‘proxy’

The following object is masked from ‘package:Matrix’:

```
as.matrix
```

The following objects are masked from ‘package:stats’:

```
as.dist, dist
```

The following object is masked from ‘package:base’:

```
as.matrix
```

Attaching package: ‘survey’

The following object is masked from ‘package:graphics’:

```
dotchart
```

Package 'GA' version 3.0.2
Type 'citation("GA")' for citing this R package in publications.
This is vegan 2.4-1

Attaching package: ‘vegan’

The following object is masked from ‘package:survey’:

```
calibrate
```

Attaching package: ‘reshape’

The following object is masked from ‘package:data.table’:

```
melt
```

The following object is masked from ‘package:Matrix’:

```
expand
```

Attaching package: ‘SparseM’

The following object is masked from ‘package:base’:

```
backsolve
```

'drc' has been loaded.

Please cite R and 'drc' if used for a publication,
for references type 'citation()' and 'citation('drc')'.

Attaching package: ‘drc’

The following object is masked from ‘package:survey’:

```
twophase
```

The following objects are masked from ‘package:stats’:

```
gaussian, getInitial
```

Attaching package: ‘gplots’

The following object is masked from ‘package:stats’:

```
lowess
```

**Comment by nlawlor** [comment]
I have also attached the gene expression file I used when attempting to run GiniClust. The file is saved as an .xlsx file because I was unable to upload the CSV version. 

[gene.expression.xlsx](https://github.com/lanjiangboston/GiniClust/files/548820/gene.expression.xlsx)

**Comment by bernard2012** [comment]
Hi nlawlor,

I am a postdoc who is one of maintainers of GiniClust.
I am able to get past your error by doing two things:
1) remove the empty space at the beginning of the first header line (the
line should start with the first column label)
2) enclose strings in "", though this step may not be essential

however, I am running into other problems that seem to be caused by the
dimension of the dataset.

As you know, the example GBM dataset was designed for full 20K gene
single-cell RNAseq dataset, while your dataset has only 2000 genes.

Do you have the full 20K version of your dataset to try GiniClust on?

There is a parameter file Giniclust_parameters.R which makes the whole tool
quite customizable.

If you want to stick with the small dataset, may be you need to change the
parameter file, specifically the following ones: minGeneNum,
expressed_cutoff, log2.expr.cutoffl, log2.expr.cutoffh. (since these relate
to the gene filtering steps)

Hope these information help.

Qian

**Comment by nlawlor** [comment]
Hi Qian,

Thank you for your reply. I do have the original matrix of roughly 15000 genes and 4000 samples, however, it was too large to upload so I shortened the matrix.

Even after I remove the empty space at the beginning of the first header line, I still receive the same error. I saw in the GiniClust_Preprocess.R function that it uses the read.csv function to read in the CSV file but does not specify the "row.names" argument. I think without this, it is creating default row names of my matrix as numbers (1, 2, 3, 4) and treating the true character row names of my matrix (Gene 1, Gene 2, etc) as part of the matrix data, hence the error: "Error in FUN(newX[, i], ...) : invalid 'type' (character) of argument".  After changing the GiniClust_Preprocess.R function to include "row.names = 1", GiniClust works well for me.

To clarify I changed this line of code:

ExprM.RawCounts  <- read.csv(data.file, sep=",", head=T)

to

ExprM.RawCounts  <- read.csv(data.file, sep=",", head=T, row.names = 1)

Thank you again for your feedback,
Nathan
